### PR TITLE
Support CreateTable RPC

### PR DESCRIPTION
**Pull request by dzh-cmd**
Add a service to create tables between clients and servers.
When the client calls this RPC, the client needs to pass in the table name and field description, and the server will store the information of the table as metadata.

### CODE REVIEW

**Comment by dzh-cmd** [comment]
I have fixed the error about "segmentation fault (core dumped)".

File: src/protobuf/rpc.proto [hunk 1]
```diff
@@ -15,3 +19,42 @@ enum CheckVersionStatus {
...
+  string name = 1;
+  FieldType type = 2;
+}
+
+message TableSchema {
+  string table_name = 1;
+  repeated Field fields = 2;
+}
+
+message CreateTableRequest { TableSchema table_schema = 1; }
```
Comment (jovany-wang) [inline src/protobuf/rpc.proto:55]: It seems that we needn't TableSchema right? We can use CreateTableRequest directly:
```proto
message CreateTableRequest {
  string table_name = 1;
  repeated Field fields = 2;
}
```

File: src/protobuf/rpc.proto [hunk 2]
```diff
@@ -15,3 +19,42 @@ enum CheckVersionStatus {
...
+
+message CreateTableReply {
+  string msg = 1;
+  CreateTableStatus status = 2;
```
Comment (jovany-wang) [inline src/protobuf/rpc.proto:59]: Maybe we always should put the status at the first of the fields?

File: src/client/metable_client.cc [hunk 3]
```diff
@@ -27,4 +27,27 @@ bool MetableClient::CheckVersion(const std::string &user) {
   }
 }
 
+std::pair<bool, std::string> MetableClient::CreateTable(
+    const std::string &name,
+    const std::vector<std::pair<std::string, rpc::FieldType>> &fields) {
+  rpc::CreateTableRequest request;
+  rpc::TableSchema *schema = new rpc::TableSchema();
```
Comment (jovany-wang) [inline src/client/metable_client.cc:34]: We shouldn't use `new` operator explicitly. These lines could be refined as:
```C++
  rpc::CreateTableRequest request;
  auto *schema = request.mutable_schema();
   ...
```

File: src/client/metable_client.h [hunk 4]
```diff
@@ -33,6 +33,16 @@ class MetableClient {
   // from the server.
   bool CheckVersion(const std::string &user);
 
+  // CreataTable RPC.
```
Comment (jovany-wang) [inline src/client/metable_client.h:36]:
```suggestion
  // Create a table on the Metable server.
```

File: src/server/metable_server.h [hunk 5]
```diff
@@ -25,6 +25,14 @@ class MetableServiceImpl : public rpc::Metable::Service {
   virtual grpc::Status CheckVersion(::grpc::ServerContext *context,
                                     const rpc::CheckVersionRequest *request,
                                     rpc::CheckVersionReply *reply) override;
+
+  virtual grpc::Status CreateTable(::grpc::ServerContext *context,
+                                   const rpc::CreateTableRequest *request,
+                                   rpc::CreateTableReply *reply) override;
+
+private:
+  // Store meta information of all tables。
```
Comment (jovany-wang) [inline src/server/metable_server.h:34]:
```suggestion
  // All tables in memeory.
```

**Comment by jovany-wang** [comment]
Could you also add your PR description?

**Review by jovany-wang — APPROVED** [review]

**Comment by jovany-wang** [comment]
Approved. Will merge it once lint fixed.

**Comment by dzh-cmd** [comment]
Resubmitted




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "jovany-wang/metable"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年3月4日(星期五) 下午5:41
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [jovany-wang/metable] Support CreateTable RPC (PR #46)





 
@jovany-wang approved this pull request.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by dzh-cmd** [comment]
>
Are there any problems with the code-format.sh script? Here are some suggestions for me. However, the indentation length of the file should be 2 instead of 4, right?
 
 message Field {
-  string name = 1;
-  FieldType type = 2;
+    string name = 1;
+    FieldType type = 2;
 }

**Comment by jovany-wang** [comment]
> Are there any problems with the code-format.sh script? Here are some suggestions for me. However, the indentation length of the file should be 2 instead of 4, right?

No, I believe it's 4.
image

**Comment by jovany-wang** [comment]
Let me merge this first and then I'll format the whole project in the next PR.

**Pull request merged**